codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -246,38 +274,38 @@ public ChannelFuture writePushPromise(ChannelHandlerContext ctx, int streamId,
 
             // Read the first fragment (possibly everything).
             Http2Flags flags = new Http2Flags().paddingPresent(padding > 0);
-            int promisedStreamIdLength = INT_FIELD_LENGTH;
-            int nonFragmentLength = promisedStreamIdLength + padding + flags.getPaddingPresenceFieldLength();
+            // INT_FIELD_LENGTH is for the length of the promisedStreamId
+            int nonFragmentLength = INT_FIELD_LENGTH + padding + flags.getPaddingPresenceFieldLength();
             int maxFragmentLength = maxFrameSize - nonFragmentLength;
             ByteBuf fragment =
-                    headerBlock.readSlice(Math.min(headerBlock.readableBytes(), maxFragmentLength));
+                    headerBlock.readSlice(Math.min(headerBlock.readableBytes(), maxFragmentLength)).retain();
 
-            flags.endOfHeaders(headerBlock.readableBytes() == 0);
+            flags.endOfHeaders(!headerBlock.isReadable());
 
             int payloadLength = fragment.readableBytes() + nonFragmentLength;
-            ByteBuf firstFrame = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
-            writeFrameHeader(firstFrame, payloadLength, PUSH_PROMISE, flags,
-                    streamId);
-
-            writePaddingLength(padding, firstFrame);
+            ByteBuf buf = ctx.alloc().buffer(PUSH_PROMISE_FRAME_HEADER_LENGTH);
+            writeFrameHeaderInternal(buf, payloadLength, PUSH_PROMISE, flags, streamId);
+            writePaddingLength(buf, padding);
 
             // Write out the promised stream ID.
-            firstFrame.writeInt(promisedStreamId);
+            buf.writeInt(promisedStreamId);
+            ctx.write(buf, promiseAggregator.newPromise());
 
             // Write the first fragment.
-            firstFrame.writeBytes(fragment);
+            ctx.write(fragment, promiseAggregator.newPromise());
 
-            // Write out the padding, if any.
-            firstFrame.writeZero(padding);
+            if (padding > 0) { // Write out the padding, if any.
+                buf = ZERO_BUFFER.slice(0, padding).retain();
```

Comment:
assigning to buf seems redundant and perhaps a little error prone, just inline into the write call below?